src/index.ts

```diff
@@ -0,0 +1,1003 @@
+        cb2();
+      });
+    }
+
+    // delete the block, and if block is in the canonical chain, delete all
+    // children as well
+    function buildDBops(cb2: any) {
+      self._delChild(
+        blockHash,
+        blockNumber,
+        inCanonical ? parentHash : null,
+        dbOps,
+        cb2
+      );
+    }
+
+    // delete all number to hash mappings for deleted block number and above
+    function deleteStaleAssignments(cb2: any) {
+      if (inCanonical) {
+        self._deleteStaleAssignments(blockNumber, parentHash, dbOps, cb2);
+      } else {
+        cb2();
+      }
+    }
+  }
+
+  _delChild(hash: Buffer, number: BN, headHash: Buffer, ops: any, cb: any) {
+    const self = this;
+
+    // delete header, body, hash to number mapping and td
+    ops.push({
+      type: "del",
+      key: headerKey(number, hash),
+      keyEncoding: "binary"
+    });
+    self.dbManager._cache.header.del(headerKey(number, hash));
+
+    ops.push({
+      type: "del",
+      key: bodyKey(number, hash),
+      keyEncoding: "binary"
+    });
+    self.dbManager._cache.body.del(bodyKey(number, hash));
+
+    ops.push({
+      type: "del",
+      key: hashToNumberKey(hash),
+      keyEncoding: "binary"
+    });
+    self.dbManager._cache.hashToNumber.del(hashToNumberKey(hash));
+
+    ops.push({
+      type: "del",
+      key: tdKey(number, hash),
+      keyEncoding: "binary"
+    });
+    self.dbManager._cache.td.del(tdKey(number, hash));
+
+    if (!headHash) {
+      return cb();
+    }
+
+    if (hash.equals(self._headHeader)) {
+      self._headHeader = headHash;
+    }
+
+    if (hash.equals(self._headBlock)) {
+      self._headBlock = headHash;
+    }
+
+    self._getCanonicalHeader(number.addn(1), (err?: any, childHeader?: any) => {
+      if (err) {
+        return cb();
+      }
+      self._delChild(
+        childHeader.hash(),
+        new BN(childHeader.number),
+        headHash,
+        ops,
+        cb
+      );
+    });
+  }
+
+  /**
+   * Iterates through blocks starting at the specified iterator head and calls
+   * the onBlock function on each block. The current location of an iterator head
+   * can be retrieved using the `getHead()`` method
+   */
+  iterator(name: string, onBlock: any, cb: any): void {
+    // ensure init completed
+    this._initLock.await(() => {
+      this._iterator(name, onBlock, cb);
+    });
+  }
+
+  _iterator(name: string, func: any, cb: any) {
+    const self = this;
+    const blockHash = self._heads[name] || self._genesis;
+    let blockNumber: any;
+    let lastBlock: any;
+
+    if (!blockHash) {
+      return cb();
+    }
+
+    self._hashToNumber(blockHash, (err?: any, number?: any) => {
+      if (err) return cb(err);
+      blockNumber = number.addn(1);
+      async.whilst(() => blockNumber, run, err =>
+        err ? cb(err) : self._saveHeads(cb)
+      );
+    });
+
+    function run(cb2: any) {
+      let block: any;
+
+      async.series([getBlock, runFunc], function(err: any) {
```

Comment:
Is this ``err`` here non-optional?

Reply:
It'll be `null` or a `error`, I do believe